_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,240 @@
+  work-offline: true
+  ids: 'com.epages:product-service'
+```
+
+Now all we need to do is annotate our improved test with `@AutoConfigureStubRunner` and use the `StubTrigger` to send the stub message in.
+The `spring-cloud-contract-stub-runner` downloads the stub jar with our contract from a maven repository and emits it into our application.
+This is done without interaction with a running broker.
+After the event is emitted we can run our assertions to make sure the correct effect can be observed.
+In our case, we assert that the correct product reference has been saved.
+Our [refactored test](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/test/java/com/epages/checkout/ProductSubscriberContractTest.java) looks like this:
+
+```java
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = CheckoutServiceApplication.class)
+@AutoConfigureStubRunner
+public class ProductSubscriberContractTest {
+
+  @Autowired
+  private StubTrigger stubTrigger;
+
+  @Autowired
+  private ProductRefRepository productRefRepository;
+
+  @Test
+  public void should_handle_product_created_event() {
+    //emit an event from contract with label "product.created.event"
+    stubTrigger.trigger("product.created.event");
+
+    then(productRefRepository.exists(9L)).isTrue();
+  }
+}
+```
+
+Basically, the test is covering the same aspects as our first version.
+But **it always picks up the latest contract definitions**.
+This can be called a provider contract test approach. The owner of the interface publishes the information about the structure of the event it emits (e.g. the contract).
+A consumer can use the contract to assert that events that follow the contract can be processed.
+
+## Benefits and Downsides
+
+The main benefit of this approach is the ability to always test against the contract published by the provider.
+Tests in the consumer of a message break if it no longer is able to process messages that follow the contract.
+
+Another big plus is the test generated on the provider side that makes sure the contract is in sync with the implementation.
+**This protection from accidentally breaking the contract is very important**.
+
+The price we need to pay for these benefits is writing contracts for our events and maintaining them.
+In a complex application with a lot of interactions, this can be a significant effort.
+
+At the moment Consumer-Driven Contracts do not seem to be possible to achieve for event-based interactions and Spring Cloud Contract.
+With this approach, the consumer would publish its expectations about the event produced by the provider. The provider would pick up these consumer contracts and run its tests against these. Spring Cloud Contract supports this for HTTP-based interactions but it is not for event-based interactions.
+
+All told, Spring Cloud Contract is a great tool to improve testing of event-based interaction that is well worth considering if your application is event-heavy.
```

Comment:
might want to link to my blog post about restdocs-wiremock where I explained our synchronous / HTTP based approach :)